File: DFS/79wordSearch.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        def dfs(word, board, i, j):
            if len(word) == 0:
                return True
            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or board[i][j] != word[0]:
                return False
            temp = board[i][j]
            board[i][j] = '#'
            res = dfs(word[1:], board, i+1, j) or dfs(word[1:], board, i-1, j) or dfs(word[1:], board, i, j+1) or  dfs(word[1:], board, i, j-1)
            board[i][j] = temp
            return res
        
        if not board:
            return False
        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(word, board, i, j):
                    return True
        return False

File: DFS/79WordSearch.cpp
class Solution {
private:
    bool result;
public:
    bool exist(vector<vector<char>>& board, string word) {
        result = false;
        for(int i=0; i<board.size(); i++){
            for(int j=0; j<board[0].size(); j++){
                if(board[i][j]==word[0]){
                    if(helper(board, i, j, word, 0)) return true;
                }
            }
        }
        return result;
    }
    bool helper(vector<vector<char>>& board, int x, int y, string word, int index){
        if(x>=0 && y>=0 && x<board.size() && y<board[0].size() && board[x][y]==word[index]){
            if(index==word.size()-1) return true;
            char c = board[x][y];
            board[x][y] = '#';
            bool b = helper(board, x+1, y, word, index+1) || helper(board, x-1, y, word, index+1) || helper(board, x, y+1, word, index+1) || helper(board, x, y-1, word, index+1);
            board[x][y] = c;
            return b;
        }else return false;
    }
};